### PR TITLE
[y_cable] refactor upgrade firmware API's; Fix vendor and part number API's read size for read_eeprom

### PR DESCRIPTION
**Pull request by vdahiya12**
This PR adds stub functions definition with details as to how firmware upgrade should be implemented by a Y cable vendor. 
It also fixes the read_eeprom read size for vendor part number and vendor name retreival API's

#### Description
Added a stub functions definition with a doc string describing how it should be implemented
Added these API's stub functions for vendors to implement
```
def download_firmware(physical_port, fwfile)
def activate_firmware(physical_port)
def rollback_firmware(physical_port)
```

#### Motivation and Context

Firmware upgrade is a functionality that vendors need to implement, this PR adds the definition's
and a description of how the implementation of this firmware upgrade  procedure API's should be implemented
by breaking down the firmware upgrade into above specified sub routines.

#### How Has This Been Tested?
Will be tested once implemented
for cleanup API's opened a python shell and executed API's for correctness

#### Additional Information (Optional)

Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

### CODE REVIEW

File: sonic_y_cable/y_cable.py [hunk 1]
```diff
@@ -1334,14 +1337,17 @@ def get_nic_voltage(physical_port):
...
-def upgrade_firmware(physical_port, fwfile):
-    """ This routine should facilitate complete firmware
-    upgrade of the Y cable on all the three ends of the
-    Y cable of the port specified.
-    All the components of the Y cable should be upgraded and committed
-    in their entirety by this single call subroutine.
-    This should return success code if firmware upgrade is successful
-    and an error code otherwise.
+def download_firmware(physical_port, fwfile):
+    """ This routine should basically download and store the firmware
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1341]: Remove 'basically'; unnecessarily verbose.

File: sonic_y_cable/y_cable.py [hunk 2]
```diff
@@ -1334,14 +1337,17 @@ def get_nic_voltage(physical_port):
...
+    of the Y cable on all the componenets of the Y cable of the port specified.
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1342]: Remove the first 'of the Y cable'; redundant.

File: sonic_y_cable/y_cable.py [hunk 3]
```diff
@@ -1334,14 +1337,17 @@ def get_nic_voltage(physical_port):
...
+    This should include any internal transfers ,checksum validation etc.
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1343]: Comma before space, not after.

File: sonic_y_cable/y_cable.py [hunk 4]
```diff
@@ -1334,14 +1337,17 @@ def get_nic_voltage(physical_port):
...
+    from TOR to TOR or TOR to NIC side of the firmware specified by the fwfile.
+    This basically means that the firmware which is being downloaded should be
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1345]: Remove 'basically'; unnecessarily verbose.

File: sonic_y_cable/y_cable.py [hunk 5]
```diff
@@ -1334,14 +1337,17 @@ def get_nic_voltage(physical_port):
...
+    available to be activated(start being utilized by the cable) once this API is
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1346]: Ad space before `(`

File: sonic_y_cable/y_cable.py [hunk 6]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
              a string, a path to the binary file which contains the firmware image
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1356]: I think we should add a note that the firmware file can be in the format of the vendor's choosing (binary, archive, etc.). But note that it should be one file which contains firmware for all components of the Y-cable.

File: sonic_y_cable/y_cable.py [hunk 7]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
...
-             a predefined code stating whether the firmware upgrade was successful
-             or an error code as to what was the cause of firmware upgrade failure
+             a predefined code stating whether the firmware download was successful
+             or an error code as to what was the cause of firmware download failure
     """
 
     return FIRMWARE_UPGRADE_SUCCESS
+
+def activate_firmware(physical_port):
+    """ This routine should basically activate the downloaded firmware
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1366]: Remove 'basically'; unnecessarily verbose.

File: sonic_y_cable/y_cable.py [hunk 8]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
...
+    of the Y cable on all the componenets of the Y cable of the port specified.
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1367]: Remove the first 'of the Y cable'; redundant.

File: sonic_y_cable/y_cable.py [hunk 9]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
...
+    This API is meant to be used in conjunction with download_firmware API, and
+    should be called once download_firmware API is succesful.
+    This basically means that the firmware which has been downloaded should be
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1370]: Remove 'basically'; unnecessarily verbose.

File: sonic_y_cable/y_cable.py [hunk 10]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
...
+    activated(start being utilized by the cable) once this API is
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1371]: Ad space before `(`

File: sonic_y_cable/y_cable.py [hunk 11]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
...
+    Returns:
+        an Integer:
+             a predefined code stating whether the firmware activate was successful
+             or an error code as to what was the cause of firmware activate failure
+    """
+
+    return FIRMWARE_ACTIVATE_SUCCESS
+
+def rollback_firmware(physical_port):
+    """ This routine should basically rollback the firmware to the previous version
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1386]: Remove 'basically'; unnecessarily verbose.

File: sonic_y_cable/y_cable.py [hunk 12]
```diff
@@ -1350,8 +1356,45 @@ def upgrade_firmware(physical_port, fwfile):
...
+    which was being used by the cable. This API should be mostly called when the
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1387]: s/should be mostly/is intended to be/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 1 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 2 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 3 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1343]: fixed

File: sonic_y_cable/y_cable.py (hunk 4 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py:1345]: fixed

File: sonic_y_cable/y_cable.py (hunk 5 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 8 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 9 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 10 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 11 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py (hunk 12 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

File: sonic_y_cable/y_cable.py [hunk 13]
```diff
@@ -1334,24 +1337,67 @@ def get_nic_voltage(physical_port):
...
+    Note that this API should ideally not require any rollback even if it fails
+    as this should not interfere with the existing cable functionality because
+    this has not been activated yet.
 
     Args:
         physical_port:
              an Integer, the actual physical port connected to a Y cable
         fwfile:
-             a string, a path to the binary file which contains the firmware image
+             a string, a path to the binary file which contains the firmware image.
```
Comment (jleveque) [inline sonic_y_cable/y_cable.py:1356]: Remove 'binary' from this line now that we have a detailed description of the file format.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: sonic_y_cable/y_cable.py (hunk 13 above)
Comment (vdahiya12) [reply sonic_y_cable/y_cable.py]: fixed

**Review by jleveque — APPROVED** [review]

**Pull request merged**